File: java/Greedy_java/Pancake_Sorting_969.java
package java.Greedy_java;

import java.util.*;
import java.io.*;

// import static java.lang.Math.max;
// import static java.lang.Math.min;
// import static java.lang.Math.abs;
// import static java.lang.System.out;

public class Pancake_Sorting_969 {

    /*
     * can be many valid ans easy but treaky
     * we start from last index and se if any max value present before that index.
     * if(yes)
     * - then add maxValueIndx+1 to ans and ithIndex aswell;
     * - then make 2 reverse operation first from 0-maxValueIndex and 2 0 to
     * -- ithIndex
     * else
     * - don't do any thing
     * 
     * 
     */
    public void reverse(int[] arr, int l, int r) {
        while (l < r) {
            arr[l] ^= arr[r];
            arr[r] ^= arr[l];
            arr[l] ^= arr[r];
            l++;
            r--;
        }
    }

    public List<Integer> pancakeSort(int[] arr) {
        List<Integer> ans = new ArrayList<Integer>();
        for (int i = arr.length - 1; i >= 0; --i) {
            int indx = i;
            for (int j = i - 1; j >= 0; --j) {
                indx = (arr[indx] < arr[j] ? j : indx);
            }
            if (indx != i) {
                ans.add(indx + 1);
                ans.add(i + 1);
                reverse(arr, 0, indx);
                reverse(arr, 0, i);
            }
        }
        return ans;
    }

    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("output.txt");

        // Scanner ip = new Scanner(System.in);
        try (Scanner ip = new Scanner(new File("input.txt"));
                PrintWriter op = new PrintWriter(f);) {
            Pancake_Sorting_969 obj = new Pancake_Sorting_969();
            int grid[] = { 3, 2, 4, 1 };
            // for above grid both are correct ans [4,2,4,3], [3, 4, 2, 3, 1, 2]
            op.println(obj.pancakeSort(grid));

        }
    }

}

File: Code/Greedy/Pancake_Sorting_969.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    /* can be many valid ans easy but treaky
        we start from last index and se if any max value present before that index.
        if(yes)
            then add maxValueIndx+1 to ans and ithIndex aswell;
            then make 2 reverse operation first from 0-maxValueIndex and 2 0 to ithIndex
        else
            don't do any thing


    */
    void reverse(vector<int> &arr, int l, int r)
    {
        while (l < r)
        {
            swap(arr[l], arr[r]);
            l++;
            r--;
        }
    }
    vector<int> pancakeSort(vector<int> arr)
    {
        vector<int> ans;
        for (int i = arr.size() - 1; i >= 0; --i)
        {
            int indx = i;
            for (int j = i - 1; j >= 0; --j)
            {
                indx = (arr[indx] < arr[j] ? j : indx);
            }
            if (indx != i)
            {
                ans.push_back(indx + 1);
                ans.push_back(i + 1);
                reverse(arr, 0, indx);
                reverse(arr, 0, i);
            }
        }
        return ans;
    }
};
int main()
{
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    Solution obj;
    auto ans = obj.pancakeSort({3, 2, 4, 1});
    // cout << ans;
    return 0;
}